File: CS/DXApplication3/CustomDescriptor/UnboundRowHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using DevExpress.XtraVerticalGrid;
using DevExpress.XtraVerticalGrid.Events;
using System.ComponentModel;
using System.Collections;
using DevExpress.XtraVerticalGrid.Rows;

namespace DXApplication3
{
    public class UnboundRowsHelper
    {


        private readonly PropertyGridControl _PropertyGrid;

        private List<UnboundRow> _UnboundRows = new List<UnboundRow>();

        public UnboundRowsHelper(PropertyGridControl pg)
        {
            _PropertyGrid = pg;
            _PropertyGrid.CustomPropertyDescriptors += _PropertyGrid_CustomPropertyDescriptors;
        }

        void _PropertyGrid_CustomPropertyDescriptors(object sender, CustomPropertyDescriptorsEventArgs e)
        {
            if ((sender as PropertyGridControl).SelectedObject == e.Source)
            {
                PropertyDescriptorCollection properties = e.Properties;
                ArrayList list = new ArrayList(properties);
                list.AddRange(_UnboundRows);
                PropertyDescriptor[] result = new PropertyDescriptor[list.Count];
                list.ToArray().CopyTo(result, 0);
                e.Properties = new PropertyDescriptorCollection(result);
            }
        }

        public void AddUnboundRow(string fieldName, Type valueType)
        {
            
            UnboundRow row = new UnboundRow(fieldName, valueType);
            _UnboundRows.Add(row);
        }
    }
}

File: VB/DXApplication3/CustomDescriptor/UnboundRowHelper.vb
Imports System
Imports System.Collections.Generic
Imports DevExpress.XtraVerticalGrid
Imports DevExpress.XtraVerticalGrid.Events
Imports System.ComponentModel
Imports System.Collections

Namespace DXApplication3

    Public Class UnboundRowsHelper

        Private ReadOnly _PropertyGrid As PropertyGridControl

        Private _UnboundRows As List(Of UnboundRow) = New List(Of UnboundRow)()

        Public Sub New(ByVal pg As PropertyGridControl)
            _PropertyGrid = pg
            AddHandler _PropertyGrid.CustomPropertyDescriptors, AddressOf _PropertyGrid_CustomPropertyDescriptors
        End Sub

        Private Sub _PropertyGrid_CustomPropertyDescriptors(ByVal sender As Object, ByVal e As CustomPropertyDescriptorsEventArgs)
            If TryCast(sender, PropertyGridControl).SelectedObject Is e.Source Then
                Dim properties As PropertyDescriptorCollection = e.Properties
                Dim list As ArrayList = New ArrayList(properties)
                list.AddRange(_UnboundRows)
                Dim result As PropertyDescriptor() = New PropertyDescriptor(list.Count - 1) {}
                list.ToArray().CopyTo(result, 0)
                e.Properties = New PropertyDescriptorCollection(result)
            End If
        End Sub

        Public Sub AddUnboundRow(ByVal fieldName As String, ByVal valueType As Type)
            Dim row As UnboundRow = New UnboundRow(fieldName, valueType)
            _UnboundRows.Add(row)
        End Sub
    End Class
End Namespace
